File: widget/app/src/main/java/com/example/structuredwidget/data/BackendClient.kt
package com.example.structuredwidget.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class BackendClient(
    private val credentials: ApiCredentials,
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .build()

    data class ProbeResult(
        val ok: Boolean,
        val message: String,
        val httpCode: Int? = null,
    )

    sealed class FetchError {
        data object Unauthorized : FetchError()
        data class Http(val code: Int, val message: String) : FetchError()
        data class Network(val message: String) : FetchError()
    }

    sealed class SnapshotFetchResult {
        data class Ok(val snapshot: WidgetSnapshot, val etag: String?) : SnapshotFetchResult()
        data class NotModified(val etag: String?) : SnapshotFetchResult()
        data class Error(val error: FetchError) : SnapshotFetchResult()
    }

    /** GET /v1/me — validates URL + Discord credentials. */
    suspend fun probeMe(): ProbeResult = withContext(Dispatchers.IO) {
        val base = credentials.getBaseUrl()
        val discordId = credentials.getDiscordId()
        val token = credentials.getWidgetToken()
        if (base.isNullOrBlank() || discordId.isNullOrBlank() || token.isNullOrBlank()) {
            return@withContext ProbeResult(false, "Missing URL, Discord ID, or token")
        }
        val url = "$base/v1/me"
        AppLog.d("GET $url (discord=$discordId)")
        try {
            val request = authRequest(url).get().build()
            client.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    AppLog.w("GET $url → ${response.code} ${body.take(160)}")
                    return@use ProbeResult(
                        ok = false,
                        message = "HTTP ${response.code}: ${body.take(100).ifBlank { response.message }}",
                        httpCode = response.code,
                    )
                }
                val tz = runCatching { JSONObject(body).optString("timezone") }.getOrNull()
                val msg = if (!tz.isNullOrBlank()) "Authenticated · timezone $tz" else "Authenticated"
                AppLog.i("GET $url → ${response.code} $msg")
                ProbeResult(true, msg, response.code)
            }
        } catch (e: kotlin.coroutines.cancellation.CancellationException) {
            throw e
        } catch (e: Exception) {
            AppLog.e("GET $url failed", e)
            ProbeResult(false, e.message ?: e.javaClass.simpleName)
        }
    }

    /** GET /v1/widget/snapshot — single atomic widget payload. */
    suspend fun fetchSnapshot(ifNoneMatch: String? = null): SnapshotFetchResult =
        withContext(Dispatchers.IO) {
            val base = credentials.getBaseUrl()
                ?: return@withContext SnapshotFetchResult.Error(FetchError.Network("Base URL not set"))
            val discordId = credentials.getDiscordId()
                ?: return@withContext SnapshotFetchResult.Error(FetchError.Network("Discord ID not set"))
            val token = credentials.getWidgetToken()
                ?: return@withContext SnapshotFetchResult.Error(FetchError.Network("Widget token not set"))
            val url = "$base/v1/widget/snapshot"
            AppLog.d("GET $url (discord=$discordId etag=${ifNoneMatch != null})")
            try {
                val builder = authRequest(url).get()
                if (!ifNoneMatch.isNullOrBlank()) {
                    builder.header("If-None-Match", ifNoneMatch)
                }
                client.newCall(builder.build()).execute().use { response ->
                    val etag = response.header("ETag")
                    when {
                        response.code == 304 ->
                            SnapshotFetchResult.NotModified(etag)
                        response.code == 401 -> {
                            AppLog.w("GET $url → 401 unauthorized")
                            SnapshotFetchResult.Error(FetchError.Unauthorized)
                        }
                        !response.isSuccessful -> {
                            val body = response.body?.string().orEmpty()
                            AppLog.w("GET $url → ${response.code} ${body.take(160)}")
                            SnapshotFetchResult.Error(
                                FetchError.Http(
                                    response.code,
                                    body.take(120).ifBlank { response.message },
                                ),
                            )
                        }
                        else -> {
                            val body = response.body?.string().orEmpty()
                            val snapshot = WidgetSnapshotParser.parse(JSONObject(body))
                            AppLog.i(
                                "GET $url → ${response.code} v=${snapshot.version} " +
                                    "logical=${snapshot.logicalDate} etag=$etag",
                            )
                            SnapshotFetchResult.Ok(snapshot, etag)
                        }
                    }
                }
            } catch (e: kotlin.coroutines.cancellation.CancellationException) {
                throw e
            } catch (e: Exception) {
                AppLog.e("GET $url failed", e)
                SnapshotFetchResult.Error(FetchError.Network(e.message ?: e.javaClass.simpleName))
            }
        }

    private fun authRequest(url: String): Request.Builder =
        Request.Builder()
            .url(url)
            .header("Accept", "application/json")
            .header("X-Discord-Id", credentials.getDiscordId().orEmpty())
            .header("X-Widget-Token", credentials.getWidgetToken().orEmpty())
}

File: ios/Sources/StructuredCore/BackendClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct HTTPRequest: Equatable, Sendable {
    public var method: String
    public var url: String
    public var headers: [String: String]

    public init(method: String, url: String, headers: [String: String] = [:]) {
        self.method = method
        self.url = url
        self.headers = headers
    }
}

public struct HTTPResponse: Equatable, Sendable {
    public var statusCode: Int
    public var headers: [String: String]
    public var body: Data

    public init(statusCode: Int, headers: [String: String] = [:], body: Data = Data()) {
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
    }

    public var headerETag: String? {
        headers.first { $0.key.lowercased() == "etag" }?.value
    }
}

public protocol HTTPTransporting: Sendable {
    func send(_ request: HTTPRequest) async throws -> HTTPResponse
}

public struct BackendClient: SnapshotFetching, MeProbing, Sendable {
    public var credentials: Credentials
    public var transport: any HTTPTransporting

    public init(credentials: Credentials, transport: any HTTPTransporting) {
        self.credentials = credentials
        self.transport = transport
    }

    public func probeMe() async -> ProbeResult {
        guard credentials.isConfigured else {
            return ProbeResult(ok: false, message: "Missing URL, Discord ID, or token")
        }
        let url = credentials.baseURL + "/v1/me"
        let request = authorizedGET(url)
        do {
            let response = try await transport.send(request)
            if !(200..<300).contains(response.statusCode) {
                let snippet = String(data: response.body, encoding: .utf8)?.prefix(100) ?? ""
                let msg = snippet.isEmpty ? "HTTP \(response.statusCode)" : String(snippet)
                return ProbeResult(
                    ok: false,
                    message: "HTTP \(response.statusCode): \(msg)",
                    httpCode: response.statusCode
                )
            }
            let tz = parseTimezone(response.body)
            let msg = tz.map { "Authenticated · timezone \($0)" } ?? "Authenticated"
            return ProbeResult(ok: true, message: msg, httpCode: response.statusCode)
        } catch {
            return ProbeResult(ok: false, message: error.localizedDescription)
        }
    }

    public func fetchSnapshot(ifNoneMatch: String? = nil) async -> SnapshotFetchResult {
        guard credentials.isConfigured else {
            return .error(.network("Base URL not set"))
        }
        let url = credentials.baseURL + "/v1/widget/snapshot"
        var request = authorizedGET(url)
        if let ifNoneMatch, !ifNoneMatch.isEmpty {
            request.headers["If-None-Match"] = ifNoneMatch
        }
        do {
            let response = try await transport.send(request)
            let etag = response.headerETag
            if response.statusCode == 304 {
                return .notModified(etag: etag)
            }
            if response.statusCode == 401 {
                return .error(.unauthorized)
            }
            if !(200..<300).contains(response.statusCode) {
                let snippet = String(data: response.body, encoding: .utf8)?.prefix(120) ?? ""
                return .error(.http(code: response.statusCode, message: String(snippet)))
            }
            do {
                let snapshot = try WidgetSnapshotParser.parse(data: response.body)
                return .ok(snapshot: snapshot, etag: etag)
            } catch {
                return .error(.network("Invalid snapshot JSON"))
            }
        } catch {
            return .error(.network(error.localizedDescription))
        }
    }

    private func authorizedGET(_ url: String) -> HTTPRequest {
        HTTPRequest(
            method: "GET",
            url: url,
            headers: [
                "Accept": "application/json",
                "X-Discord-Id": credentials.discordId,
                "X-Widget-Token": credentials.widgetToken,
            ]
        )
    }

    private func parseTimezone(_ data: Data) -> String? {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tz = obj["timezone"] as? String,
              !tz.isEmpty
        else { return nil }
        return tz
    }
}

public struct URLSessionTransport: HTTPTransporting, @unchecked Sendable {
    public var timeout: TimeInterval

    public init(timeout: TimeInterval = 20) {
        self.timeout = timeout
    }

    public func send(_ request: HTTPRequest) async throws -> HTTPResponse {
        guard let url = URL(string: request.url) else {
            throw URLError(.badURL)
        }
        var urlRequest = URLRequest(url: url, timeoutInterval: timeout)
        urlRequest.httpMethod = request.method
        for (key, value) in request.headers {
            urlRequest.setValue(value, forHTTPHeaderField: key)
        }
        let (data, response) = try await urlSessionData(urlRequest)
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        var headers: [String: String] = [:]
        for (key, value) in http.allHeaderFields {
            if let k = key as? String, let v = value as? String {
                headers[k] = v
            }
        }
        return HTTPResponse(statusCode: http.statusCode, headers: headers, body: data)
    }
}

private func urlSessionData(_ request: URLRequest) async throws -> (Data, URLResponse) {
    #if canImport(FoundationNetworking)
    try await withCheckedThrowingContinuation { continuation in
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error {
                continuation.resume(throwing: error)
                return
            }
            guard let data, let response else {
                continuation.resume(throwing: URLError(.badServerResponse))
                return
            }
            continuation.resume(returning: (data, response))
        }
        task.resume()
    }
    #else
    try await URLSession.shared.data(for: request)
    #endif
}
